packages/network/src/logic/node/ProxyStreamConnectionManager.ts

```diff
@@ -0,0 +1,211 @@
+import { TrackerManager } from './TrackerManager'
+import { StreamManager } from './StreamManager'
+import { NodeToNode } from '../../protocol/NodeToNode'
+import { Event, Node, NodeId } from './Node'
+import {
+    PublishStreamConnectionRequest,
+    PublishStreamConnectionResponse,
+    SPID,
+    SPIDKey,
+    UnsubscribeRequest
+} from 'streamr-client-protocol'
+import { promiseTimeout } from '../../helpers/PromiseTools'
+import { Logger } from '../../helpers/logger/LoggerNode'
+const logger = new Logger(module)
+
+export interface ProxyStreamConnectionManagerOptions {
+    trackerManager: TrackerManager,
+    streamManager: StreamManager,
+    nodeToNode: NodeToNode,
+    node: Node,
+    nodeConnectTimeout: number,
+    acceptProxyConnections: boolean
+}
+
+enum State {
+    CONNECTING,
+    CONNECTED,
+    RECONNECTING
+}
+
+interface ProxyConnection {
+    state?: State,
+    reconnectionTimer?: NodeJS.Timeout
+}
+
+const DEFAULT_RECONNECTION_TIMEOUT = 10 * 1000
+
+export class ProxyStreamConnectionManager {
+    private readonly trackerManager: TrackerManager
+    private readonly streamManager: StreamManager
+    private readonly nodeToNode: NodeToNode
+    private readonly node: Node
+    private readonly nodeConnectTimeout: number
+    private readonly acceptProxyConnections: boolean
+    private readonly connections: Map<SPIDKey, Map<NodeId, ProxyConnection>>
+
+    constructor(opts: ProxyStreamConnectionManagerOptions) {
+        this.trackerManager = opts.trackerManager
+        this.streamManager = opts.streamManager
+        this.nodeToNode = opts.nodeToNode
+        this.node = opts.node
+        this.nodeConnectTimeout = opts.nodeConnectTimeout
+        this.acceptProxyConnections = opts.acceptProxyConnections
+        this.connections = new Map()
+    }
+
+    addConnection(spid: SPID, nodeId: NodeId): void {
```

Comment:
Using stricter access modifiers when possible enhances the reader's ability to reason about the code. (E.g. when used here, I know that this method is only invoked within this class.)
```suggestion
    private addConnection(spid: SPID, nodeId: NodeId): void {
```